src/audio/eq_iir.c

```diff
@@ -325,90 +342,111 @@ static int eq_iir_params(struct comp_dev *dev)
 	struct comp_data *cd = comp_get_drvdata(dev);
-	int j;
-	uint32_t ch;
-	bool val;
-
-	if (cdata->cmd == SOF_CTRL_CMD_SWITCH) {
-		trace_eq_iir("mst");
-		for (j = 0; j < cdata->num_elems; j++) {
-			ch = cdata->chanv[j].channel;
-			val = cdata->chanv[j].value;
-			tracev_value(ch);
-			tracev_value(val);
-			if (ch >= PLATFORM_MAX_CHANNELS) {
-				trace_eq_iir_error("che");
-				return -EINVAL;
-			}
-			if (val)
-				iir_unmute_df2t(&cd->iir[ch]);
-			else
-				iir_mute_df2t(&cd->iir[ch]);
-		}
-	} else {
-		trace_eq_iir_error("ste");
-		return -EINVAL;
-	}
 
-	return 0;
+	size_t bs;
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		trace_eq("gbi");
+
+		/* Copy back to user space */
+		bs = cd->config->size;
+		if (bs > SOF_EQ_IIR_MAX_SIZE || bs < 1)
+			return -EINVAL;
+		if (!cd->config)
+			memcpy(cdata->data->data, cd->config, bs);
+		break;
+	default:
+		trace_eq_error("egt");
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
 }
 
-static int iir_cmd_set_data(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+static int iir_cmd_set_data(struct comp_dev *dev,
+			    struct sof_ipc_ctrl_data *cdata)
 {
 	struct comp_data *cd = comp_get_drvdata(dev);
 	struct sof_ipc_ctrl_value_comp *compv;
+	struct sof_eq_iir_config *cfg;
+	size_t bs;
 	int i;
 	int ret = 0;
-	size_t bs;
+
+	/* Check version from ABI header */
+	if (cdata->data->comp_abi != SOF_EQ_IIR_ABI_VERSION) {
+		trace_eq_error("eab");
+		return -EINVAL;
+	}
 
 	switch (cdata->cmd) {
 	case SOF_CTRL_CMD_ENUM:
-		trace_eq_iir("EIe");
+		trace_eq("snu");
+		compv = (struct sof_ipc_ctrl_value_comp *)cdata->data->data;
 		if (cdata->index == SOF_EQ_IIR_IDX_SWITCH) {
-			trace_eq_iir("EIs");
-			compv = (struct sof_ipc_ctrl_value_comp *) cdata->data->data;
-			for (i = 0; i < (int) cdata->num_elems; i++) {
+			trace_eq("isw");
+			for (i = 0; i < (int)cdata->num_elems; i++) {
 				tracev_value(compv[i].index);
 				tracev_value(compv[i].svalue);
-				ret = eq_iir_switch_response(cd->iir, cd->config,
-					compv[i].index, compv[i].svalue);
+				ret = eq_iir_switch_response(cd->iir,
+							     cd->config,
+							     compv[i].index,
+							     compv[i].svalue);
 				if (ret < 0) {
-					trace_eq_iir_error("swe");
+					trace_eq_error("esw");
 					return -EINVAL;
 				}
 			}
 		} else {
-			trace_eq_iir_error("une");
+			trace_eq_error("enu");
 			trace_error_value(cdata->index);
 			return -EINVAL;
 		}
 		break;
 	case SOF_CTRL_CMD_BINARY:
-		trace_eq_iir("EIb");
+		trace_eq("sbi");
 		/* Check and free old config */
 		eq_iir_free_parameters(&cd->config);
 
-		/* Copy new config, need to decode data to know the size */
-		bs = cdata->data->size;
-		if ((bs > SOF_EQ_IIR_MAX_SIZE) || (bs < 1))
+		/* Copy new config, find size from header */
+		if (!cdata->data->data) {
+			trace_eq_error("edn");
+			return -EINVAL;
+		}
+
+		cfg = (struct sof_eq_iir_config *)cdata->data->data;
+		bs = cfg->size;
+		trace_value(bs);
+		if (bs > SOF_EQ_IIR_MAX_SIZE || bs < 1)
```

Comment:
|| bs == 0) ?

Reply:
Yes, here too the unnecessary caution for signed